common/src/main/java/com/github/khanshoaib3/minecraft_access/features/inventory_controls/GroupGenerator.java

```diff
@@ -467,8 +466,51 @@ else if (screen.getHandler() instanceof HopperScreenHandler)
             foundGroups.add(itemOutputGroup);
         }
 
-        if (unknownGroup.slotItems.size() > 0) {
-            foundGroups.add(unknownGroup);
+        // Adjacency is used here instead of slot.inventory because some mods have non-adjacent slots in the same
+        // inventory and adjacent slots in different inventories
+        List<List<SlotItem>> unknownGroups = new ArrayList<>(unknownSlots.size());
+        for (SlotItem slot : unknownSlots) {
+            unknownGroups.stream()
+                    .filter(group ->
+                            group.stream()
+                                    .anyMatch(groupSlot ->
+                                            groupSlot.x == slot.x && (groupSlot.y == slot.y + 18 || groupSlot.y == slot.y - 18)
+                                            || groupSlot.y == slot.y && (groupSlot.x == slot.x + 18 || groupSlot.x == slot.x - 18)
+                                    )
+                    )
+                    .findFirst()
+                    .or(() -> {
+                        List<SlotItem> group = new ArrayList<>(unknownSlots.size());
+                        unknownGroups.add(group);
+                        return Optional.of(group);
+                    })
```

Comment:
what's meaning of logic inside this or operation?
It seems like the result is always true since `new ArrayList<>(unknownSlots.size())` will never be null